core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -20,33 +20,143 @@
+  private CachingCatalog(Catalog catalog, boolean caseSensitive, boolean isExpirationEnabled,
+      long expirationIntervalInMillis) {
+    this(catalog, caseSensitive, isExpirationEnabled, expirationIntervalInMillis, Ticker.systemTicker());
+  }
+
+  private CachingCatalog(Catalog catalog, boolean caseSensitive, boolean isExpirationEnabled,
+      long expirationIntervalMillis, Ticker ticker) {
     this.catalog = catalog;
     this.caseSensitive = caseSensitive;
+    this.expirationEnabled = isExpirationEnabled;
+    this.expirationIntervalMillis = expirationIntervalMillis;
+
+    this.tableCache = createTableCache(ticker);
+  }
+
+  /**
+   * Return the age of an entry in the cache.
+   * <p>
+   * This method is only visible for testing the cache expiration policy, as cache invalidation is handled
+   * by the catalog and not the cache itself.
+   * <p>
+   * Returns the age of the cache entry corresponding to the identifier,  or {@code Optional.empty} if the table
+   * is not present in the cache or if no expireAfterAccess policy is present in this CachingCatalog.
+   */
+  @VisibleForTesting
+  Optional<Duration> cachedEntryAge(TableIdentifier identifier) {
+    return tableCache.policy()
+        .expireAfterAccess()
+        .flatMap(tableExpiration -> tableExpiration.ageOf(identifier));
+  }
+
+  // Returns the cached Table entry corresponding to the given identifier iff
+  // it's in the cache. Grabs the table in a way that doesn't count as an access
+  // and thus won't affect the cached entry's ttl (if enabled).
+  @VisibleForTesting
+  Optional<Table> tableFromCacheQuietly(TableIdentifier identifier) {
+    // Ensure async cleanup actions have happened.
+    tableCache.cleanUp();
+    return Optional.ofNullable(tableCache.policy().getIfPresentQuietly(identifier));
+  }
+
+  @VisibleForTesting
+  Cache<TableIdentifier, Table> cache() {
+    return tableCache;
+  }
+
+  @VisibleForTesting
+  Optional<Duration> getTimeToTTL(TableIdentifier identifier) {
+    return tableCache
+        .policy()
+        .expireAfterAccess()  // Assumes expireAfterAccess, which is what we set at cache level.
+        .flatMap(tableExpiration -> tableExpiration.ageOf(identifier)) // Get the time the table has been cached.
+        .map(age -> Duration.ofMillis(expirationIntervalMillis).minus(age));
+  }
+
+  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
+    Caffeine<TableIdentifier, Table> cacheBuilder = Caffeine
+        .newBuilder()
+        .softValues()
+        .removalListener(keyLoggingRemovalListener)
+        .writer(new CacheWriter<TableIdentifier, Table>() {
+          @Override
+          // TODO - Consider expiring and syncing any metadata tables that have a different snapshotId
+          //        upon write.
+          public void write(TableIdentifier tableIdentifier, Table table) {
+            LOG.info("Table {} was written to the catalog with snapshotId {}", tableIdentifier,
```

Comment:
This is just a notification?

Doesn't need to be info. Maybe don't log it at all?